proto/substrait/extended_expression.proto

```diff
@@ -0,0 +1,49 @@
+// SPDX-License-Identifier: Apache-2.0
+syntax = "proto3";
+
+package substrait;
+
+import "substrait/algebra.proto";
+import "substrait/extensions/extensions.proto";
+
+option csharp_namespace = "Substrait.Protobuf";
+option go_package = "github.com/substrait-io/substrait-go/proto";
+option java_multiple_files = true;
+option java_package = "io.substrait.proto";
+
+message ExpressionReference {
+  oneof expr_type {
+    Expression expression = 1;
+    AggregateFunction measure = 2;
+  }
+  NamedStruct base_schema = 2;
+  // Field names in depth-first order
+  string name = 3;
+}
+
+// Describe a set of operations to complete.
+// For compactness sake, identifiers are normalized at the plan level.
+message ExtendedExpression {
+  // Substrait version of the expression. Optional up to 0.17.0, required for later
+  // versions.
+  Version version = 6;
+
+  // a list of yaml specifications this expression may depend on
+  repeated substrait.extensions.SimpleExtensionURI extension_uris = 1;
+
+  // a list of extensions this expression may depend on
+  repeated substrait.extensions.SimpleExtensionDeclaration extensions = 2;
+
+  // one or more expression trees.
+  repeated ExpressionReference referred_expr = 3;
```

Comment:
What does it mean if there are multiple expressions?

Does the order matter?  Are these expressions to be evaluated one by one from top to bottom?
Can expressions refer to the value calculated in previous expressions (perhaps as a special sort of reference)?

Reply:
We translate the multi expressions from frontend framework, following the order it occurs in original plan node. But it doesn't requires the consume side to handle it strictly in previous order. Backend can evaluated it one by one OR evaluated the expressions that be referred by other expressions first. For example, 2 expression `a+b+c, a+b` are translated in such order, but expression `a+b` can be evaluated first and then can be reused in evaluation of `a+b+c`. Backend can analyze these expressions and do such optimization correspondingly.

Reply:
I add some comments in the doc.

Reply:
Could one expression explicitly reference another expression?  For example, if you added a new "expression reference" message you could have something like:

```
exprs[0] = a + b
exprs[1] = $exprs[0] + c
```

Reply:
I think for current substrait definition, there is no suitable semantic to represent this reference. Correct me if I am wrong. And I personally don't prefer this way. In original expressions under single rel, like project plan node, they don't explicitly track this reference too. The reuse is often done through CSE(common sub expression) optimization, may be various from different frontend and backend engines.